File: android/app/src/main/java/app/opensleep/domain/sonar/FftSonarConsumer.kt
package app.opensleep.domain.sonar

import app.opensleep.domain.UrbandroidFFT
import app.opensleep.domain.dsp.IirFilter
import app.opensleep.domain.dsp.IirFilterDesignFisher
import app.opensleep.domain.dsp.FilterPassType
import app.opensleep.domain.dsp.FilterCharacteristicsType
import kotlin.math.abs
import kotlin.math.sqrt

class FftSonarConsumer(val sampleRate: Int) {
    private var iirFilter: IirFilter? = null
    private var prevMags: FloatArray? = null

    init {
        reset()
    }

    data class ConsumerResult(val signal: Double, val activity: Float)

    fun processAndGetResult(fArr: FloatArray): ConsumerResult {
        // High-pass Butterworth filter at 17,000 Hz (order 18)
        val filtered = FloatArray(fArr.size)
        val filter = iirFilter
        var signalSum = 0.0
        if (filter != null) {
            for (i in fArr.indices) {
                val fVal = filter.step(fArr[i].toDouble()).toFloat()
                filtered[i] = fVal
                signalSum += abs(fVal)
            }
        } else {
            for (i in fArr.indices) {
                filtered[i] = fArr[i]
                signalSum += abs(fArr[i])
            }
        }

        // Run Real FFT
        val fftData = filtered.copyOf()
        UrbandroidFFT.realForward(fftData)

        // Bins for 20 kHz to 22 kHz
        val binFrom = (20000.0 * fArr.size / sampleRate).toInt().coerceIn(0, fArr.size / 2 - 1)
        val binTo = (22000.0 * fArr.size / sampleRate).toInt().coerceIn(0, fArr.size / 2 - 1)

        val currentMags = FloatArray(binTo - binFrom + 1)
        for (k in binFrom..binTo) {
            // In JTransforms layout, for k in 1..(n/2-1):
            // Re[k] is at 2*k, Im[k] is at 2*k+1
            val re = fftData[2 * k]
            val im = fftData[2 * k + 1]
            currentMags[k - binFrom] = sqrt(re * re + im * im)
        }

        var activity = 0.0f
        val prev = prevMags
        if (prev != null && prev.size == currentMags.size) {
            for (i in currentMags.indices) {
                activity += abs(currentMags[i] - prev[i])
            }
        }

        prevMags = currentMags
        return ConsumerResult(signalSum, activity)
    }

    fun reset() {
        prevMags = null
        val coefs = IirFilterDesignFisher.design(
            FilterPassType.highpass,
            FilterCharacteristicsType.butterworth,
            18,
            -1.0,
            17000.0 / sampleRate,
            -1.0
        )
        iirFilter = IirFilter(coefs)
    }
}

File: ios/opensleep/opensleep/Services/Sonar/FftSonarConsumer.swift
import Foundation

class FftSonarConsumer {
    let sampleRate: Int
    private var iirFilter: IirFilter?
    private var prevMags: [Float]?

    init(sampleRate: Int) {
        self.sampleRate = sampleRate
        let fcf1 = 17000.0 / Double(sampleRate)
        if fcf1 > 0.0 && fcf1 < 0.5 {
            let coefs = IirFilterDesignFisher.design(
                filterPassType: .highpass,
                filterCharacteristicsType: .butterworth,
                order: 18,
                ripple: -1.0,
                fcf1: fcf1,
                fcf2: -1.0
            )
            self.iirFilter = IirFilter(coefficients: coefs)
        } else {
            print("[SleepTracker] Warning: Invalid fcf1 (\(fcf1)) for sampleRate \(sampleRate). Sonar filter disabled.")
            self.iirFilter = nil
        }
    }

    struct ConsumerResult {
        let signal: Double
        let activity: Float
    }

    func processAndGetResult(_ fArr: [Float]) -> ConsumerResult {
        // High-pass Butterworth filter at 17,000 Hz (order 18)
        var filtered = [Float](repeating: 0.0, count: fArr.count)
        var signalSum = 0.0
        if let filter = iirFilter {
            for i in 0..<fArr.count {
                let fVal = Float(filter.step(Double(fArr[i])))
                filtered[i] = fVal
                signalSum += Double(abs(fVal))
            }
        } else {
            for i in 0..<fArr.count {
                filtered[i] = fArr[i]
                signalSum += Double(abs(fArr[i]))
            }
        }

        // Run Real FFT and compute power spectrum
        let power = UrbandroidFFT.powerSpectrum(filtered)

        // Extract magnitudes for bins between 20,000 Hz and 22,000 Hz
        let n = fArr.count
        let binFrom = max(0, Int(20000.0 * Double(n) / Double(sampleRate)))
        let binTo = min(power.count - 1, Int(22000.0 * Double(n) / Double(sampleRate)))

        var currentMags = [Float](repeating: 0.0, count: max(0, binTo - binFrom + 1))
        if binFrom <= binTo && currentMags.count > 0 {
            for k in binFrom...binTo {
                currentMags[k - binFrom] = sqrt(power[k])
            }
        }

        var activity: Float = 0.0
        if let prev = prevMags, prev.count == currentMags.count {
            for i in 0..<currentMags.count {
                activity += abs(currentMags[i] - prev[i])
            }
        }

        prevMags = currentMags
        return ConsumerResult(signal: signalSum, activity: activity)
    }

    func reset() {
        prevMags = nil
        let fcf1 = 17000.0 / Double(sampleRate)
        if fcf1 > 0.0 && fcf1 < 0.5 {
            let coefs = IirFilterDesignFisher.design(
                filterPassType: .highpass,
                filterCharacteristicsType: .butterworth,
                order: 18,
                ripple: -1.0,
                fcf1: fcf1,
                fcf2: -1.0
            )
            self.iirFilter = IirFilter(coefficients: coefs)
        } else {
            print("[SleepTracker] Warning: Invalid fcf1 (\(fcf1)) for sampleRate \(sampleRate). Sonar filter disabled.")
            self.iirFilter = nil
        }
    }
}
